src/ql/com/topology.cc

```
@@ -685,42 +689,73 @@ utils::UInt Topology::get_distance(Qubit source, Qubit target) const {
+ * The minimum number of hops is always >= distance(from, to).
  *
- * We assume below that a valid path exists with distance+1 hops; this fails
- * when not all qubits in a core support connections to all other cores.
- * See the check in initialization of neighbors.
+ * Inside one core, minimum number of hops == distance.
+ *
+ * Between qubits in different cores, the minimum number of hops >= distance + 1.
+ * This is because an inter-core hop cannot execute a 2q gate, so
+ * when the minimum number of hops are all inter-core hops (i.e. distance == core_distance),
+ * and no 2q gate has been placed yet, then at least one additional intra-core hop is needed for the 2q gate.
+ *
+ * But one additional hop is not sufficient in the special case when both source and target are communication qubits,
+ * and they are the only communication qubits in each core.
+ * The only option there is adding two hops in one of the cores:
+ * one to just another qubit and one back to the single communication qubit.
+ *
+ * Notice that:
+ * - When there is only one communication qubit per core, and whether source or target are normal qubits,
+ *   then the hop between the normal qubit and the communication qubit in its core can host the 2q gate.
+ * - And when the number of communication qubits per core is at least 2, and both operands are communication qubits,
+ *   then a hop between one of the operands and another communication qubit in its core can host the 2q gate.
  */
 utils::UInt Topology::get_min_hops(Qubit source, Qubit target) const {
-    utils::UInt d = get_distance(source, target);
-    utils::UInt cd = get_core_distance(source, target);
-    QL_ASSERT(cd <= d);
-    if (cd == d) {
-        return d+2;
+    if (source == target) {
+        return 0;
+    }
+    utils::UInt distance = get_distance(source, target);
+    utils::UInt core_distance = get_core_distance(source, target);
+    QL_ASSERT(core_distance <= distance);
+    if (core_distance == distance) {
```

at this point, distance >= 1 and core_distance is only 0 or 1
so this "if" implies `core_distance == distance == 1` and source and target being in different cores, therefore
`is_comm_qubit(source)` and `is_comm_qubit(target)` are always true, no?

I would rewrite as such, which to me is more readable:

```c++
if (!are_in_same_core(source, target) && is_comm_qubit(source) && is_comm_qubit(target)) {
  QL_ASSERT(distance == 1);
  if (num_comm_qubits == 1) {
    return 3;
  }
  return 2;
}
```
and then `core_distance` variable is useless.

Btw I think we should throw an exception when "edges" key is present or "connectivitiy" is "specified" and num_cores >= 2, in light of what Roberto told me about the assumption with multi-core connectivity.

---

Good. I've updated the function. I've added a comment because, despite the function's documentation, that check can be quite ilegible in some time (so any extra help to the future reader is welcome). Also, I prefer sticking to using `distance + 2` and `distance + 1` instead of `3` and `2` respectively (those `2` and `1` are also magic numbers, but more readable than `3` and `2`, IMO).

---

I have added a `connectivity == SPECIFIED && num_cores >= 2)` check, plus exception, at the `Topology` constructor, right after the `Handle connectivity key` section.